### ISSUE TITLE
Rewrite the Alda REPL as part of the client

### ISSUE THREAD

**Issue by daveyarwood** [alda-lang/alda-client-java#1]
Moved from https://github.com/alda-lang/alda/issues/154.

>Right now, `alda repl` just uses Java-Clojure interop to run `(alda.repl/start-repl!)` which means you have to wait for the Clojure runtime to load, the MIDI system to initialize, and a new Alda score to be initialized.
>
>The REPL is really just an interactive version of the Alda client, so now that we have a client, I think we should re-implement the Alda REPL in terms of the client. This would carry two major benefits:
>
>1) Immediate start-up, no waiting for the Clojure runtime.
>2) You would be interactively editing Alda scores running on existing servers. This would give you the flexibility to go back and forth between editing scores on the command-line (in your editor, etc.) and hopping into the Alda REPL to edit the same score.
>
>Hopefully this won't be too difficult. The Alda client already does most of the things we need the REPL to do, and most of them are functions in the AldaServer class. Basically we'll just be translating the JLine interface into Java and have it call functions in the AldaServer class.

**Comment by jgkamat** [comment]
HI @daveyarwood, I think I want to start work on this issue (I've seen it open for a *very* long time now, and It's been bugging me that the repl is still *so* slow). It seems like a few other things are blocked on this so it seems pretty important.

I had a few more questions about the alda side of the repl though (I think I can handle the JLine part).

1. How should the repl give the data to the server? The simplest way would be to use `AldaServer.play`, which [I tried out here](https://github.com/jgkamat/alda-client-java/commit/bc448fe57de195947979bada8c5eba19eb321217), but that didn't seem to keep state. for example, I can't run `piano:` and then `c d e`, without an error. Previously an `append` command existed which would have been perfect now but I think it was removed.
2. How should special commands be handled (for example `:load` and `:new`). I don't see anything that helps much in `AldaServer`, but I haven't looked at the clojure side of the server at all (and I don't know enough) to figure that out right now. I'm expecting the `:` commands would be done last.
3. Is there anything else I need to know (or any other features I should keep in mind before starting this)? Right now I only know about the `:` commands and playing things normally in the repl.
4. When I run `server.play` ther server prints out some debug info (like parsing/evaluating and playing). Is there a way to disable this (or should I be using a different method?)

I'm not sure if I need to peek into the clojure server at all for this, but I could if that would be needed.

Thanks again for creating alda! :smile:

**Comment by jgkamat** [comment]
Also, another question (but completely off topic): I'm really missing out `alda append`. `alda-mode` has a feature which allows loading a specific section, but it's slightly broken without this command. When the append command existed, we could 'append' the desired instrument, tempo, etc, and then we could play line by line or even note by note (which is great for debugging). [Is there any replacement or an alternate way of doing this from the alda command line](https://github.com/jgkamat/alda-mode/commit/e78fc1047ac613cd3f8109f11c048bc0c99b66b0) (preferably not from the repl)?

**Comment by daveyarwood** [comment]
Hey @jgkamat, thanks for offering to help with this! It would be great to have a REPL with a faster start-up. I'm happy to help with any questions that come up.

> 1. How should the repl give the data to the server? The simplest way would be to use `AldaServer.play`, which [I tried out here](https://github.com/jgkamat/alda-client-java/commit/bc448fe57de195947979bada8c5eba19eb321217), but that didn't seem to keep state. for example, I can't run `piano:` and then `c d e`, without an error. Previously an `append` command existed which would have been perfect now but I think it was removed.

I ended up removing the `append` command in favor of a more stateless system where an Alda worker process is an ephemeral thing that does not keep track of the state of a score-in-progress. This is ideal for a few reasons: 

- It keeps the server process light -- its only job is to keep track of worker processes and route requests and responses back and forth between clients and workers.

- A worker process only needs to keep track of one score at a time. Previously, there were some problems related to a single Alda server trying to play multiple scores at the same time.

  This is also good for memory reasons -- we don't need to worry about a worker process holding too many scores in memory and running out of memory and dying.

- If a worker process dies for any reason, the request can be retried with the next available worker. We aren't relying on any long-running process to keep track of the state of the score for us.

Of course, one consequence of this setup is that we need to rethink how to get `append`-like behavior for the purposes of things like the Alda REPL and the Emacs alda-mode. The server and workers are no longer responsible for maintaining the state of a score, so it is now up to the client.

What I'd like to do is allow the client to pass some additional context when making a `play` command. Right now, a client request may include [these options](https://github.com/alda-lang/alda-client-java/blob/master/src/alda/AldaRequestOptions.java) which are serialized into the JSON string sent to the server. I'm thinking we could add a `history` option, which could just be a string of Alda code that has already been evaluated up to this point.

Then, we could have the worker process look for this option on each request from the client and when it exists, it will evaluate `history` first, building the score as it normally does, except that it does not add any notes to its set of events. Then, while it has that score in memory, it can evaluate the body of the request, which is the code that we're appending to the score. The resulting score's event set will only contain the new notes from the body of the request.

> How should special commands be handled (for example :load and :new). I don't see anything that helps much in AldaServer, but I haven't looked at the clojure side of the server at all (and I don't know enough) to figure that out right now. I'm expecting the : commands would be done last.

Once what I describe above is implemented, `:load` and `:new` would just modify the `history` string (managed in memory in the Java REPL process) that we send to the server with each request. `:load` would read a file and overwrite `history` with its contents, and `:new` would overwrite `history` to `""`.

> Also, another question (but completely off topic): I'm really missing out `alda append`. `alda-mode` has a feature which allows loading a specific section, but it's slightly broken without this command. When the append command existed, we could 'append' the desired instrument, tempo, etc, and then we could play line by line or even note by note (which is great for debugging). [Is there any replacement or an alternate way of doing this from the alda command line](https://github.com/jgkamat/alda-mode/commit/e78fc1047ac613cd3f8109f11c048bc0c99b66b0) (preferably not from the repl)?

We could also add a `--history` option when making requests at the command line. Requests could look something like:

```bash
$ alda play --history "piano: c8 d e" --code "f g a b > c"
```

or, if you're maintaining the WIP score in a file, you could do something like:

```bash
$ alda play --history "$(cat my-score.alda)" --code "f g a b > c"
```

> When I run server.play ther server prints out some debug info (like parsing/evaluating and playing). Is there a way to disable this (or should I be using a different method?)

You could always add `/dev/null` onto the end to discard the output, i.e. `alda play -c 'piano: c' >/dev/null`. But I wouldn't be opposed to adding a `-q`/`--quiet` option to the client, as a convenience.

**Comment by jgkamat** [comment]
Thanks for clearing that up @daveyarwood! 

Removing `append` in favor of `--history` actually makes a lot of sense and it sounds like a good idea!(although I might be worried about what could happen if the server was on another machine and the score grew large). I think I'm pretty confident I can finish all of the java side now, but I'm not sure how the `history` option in `AldaRequestOptions` would be added (and If I should take a stab at that at all). [I'm able to trace the path of alda requests to here](https://github.com/alda-lang/alda-client-java/blob/master/src/alda/AldaRequest.java#L80), but I'm pretty confused as to where they go next (I still don't know clojure too well :stuck_out_tongue:). Would this require modification in the alda server core itself (as in parsing/evaluating in the core), or is that already done (and we just need to add a field to both sides of the interface)?

One more question: Is `*current-score*` the variable that's storing the history in the [alda repl right now](https://github.com/alda-lang/alda-repl-clj/blob/c7e1ac16999fe4c633f92b20bfc80449ed4072c1/src/alda/repl.clj)? (again, I'm pretty new at clojure so I'm not sure exactly whats going on).

**Comment by daveyarwood** [comment]
These are all great questions. I realize this is a bit complicated, but I'll do my best to walk you through how this all works. Feel free to keep asking questions whenever you have them!

> (although I might be worried about what could happen if the server was on another machine and the score grew large)

I think the biggest potential consequence is that for large scores, it might take up to a matter of seconds with each subsequent line of code entered into the REPL, for the score to be re-evaluated and the new notes played by the server. This is why my first idea (the one used by the current Clojure implementation of the Alda REPL) was to hold the score state in memory on the server so that we only need to parse and evaluate each line of code as it is entered into the REPL.

I'm not too worried about that possibility, though. I think that most of the time in cases where we're using this history functionality (i.e. in an Alda REPL or Emacs buffer), users will be evaluating short scores / code snippets, which will be very fast. Evaluating an entire score will take longer to kick off, which I think is to be expected.

Another thing that will hopefully help is that I'm currently working on a [newer, faster iteration of the parser](https://github.com/alda-lang/alda-core/issues/20), which is our major bottleneck in the time it takes to play a large Alda score. I still have a lot of work to do, but it already seems like this will speed up parsing significantly, which could make re-evaluating the entire score each time less of a problem.

> [I'm able to trace the path of alda requests to here](https://github.com/alda-lang/alda-client-java/blob/master/src/alda/AldaRequest.java#L80), but I'm pretty confused as to where they go next

The options are actually already added to the request before it is JSONified on [this line](https://github.com/alda-lang/alda-client-java/blob/master/src/alda/AldaRequest.java#L75).

The place to add the `history` option would be [here](https://github.com/alda-lang/alda-client-java/blob/master/src/alda/AldaServer.java#L247-L253) when the `play` request is constructed based on the CLI flags.

> Would this require modification in the alda server core itself (as in parsing/evaluating in the core), or is that already done (and we just need to add a field to both sides of the interface)?

We'll need to make some changes to the [worker code](https://github.com/alda-lang/alda-server-clj/blob/master/src/alda/worker.clj) in alda-server-clj. Basically, we want to look for the `history` option and if it exists, do the following:

- Parse the `history` string of Alda code, returning an error if an exception is thrown in the process. (This is what we currently do with the body of the request.)

- If this is successful, the return value is a "score map," i.e. an Alda score in data form, ready to be played. This map includes a set of events as the value of the `:events` key -- these are the notes that will be played when the score is played. So, we want to empty the `:events`.

- Then, we want to parse the body of the request [as events instead of a new score](https://github.com/alda-lang/alda-core/blob/master/src/alda/parser.clj#L253-L254), and [continue](https://github.com/alda-lang/alda-core/blob/master/src/alda/lisp/score.clj#L44-L48) the score from before with the new events.

From there, we should have a score that we can use to play only the new events.

As an aside, you might find [this blog post](http://blog.djy.io/alda-for-clojurists/) a good place to start if you're curious to know more about how the Clojure parts of Alda work.

> One more question: Is `*current-score*` the variable that's storing the history in the [alda repl right now](https://github.com/alda-lang/alda-repl-clj/blob/c7e1ac16999fe4c633f92b20bfc80449ed4072c1/src/alda/repl.clj)? (again, I'm pretty new at clojure so I'm not sure exactly whats going on). 

It is! I'm not sure how useful that code is for this project, though. It works a bit differently, as it is all in-memory in a single process. We have to do things a bit differently, with our current multi-process setup.

For now, I think a PR to add the `--history` option to `play` requests would be a good start, and we can work upward from there.

**Comment by jgkamat** [comment]
@daveyarwood this is probably going to take a couple of weeks (there's a lot of stuff to move over, and I want to do it as well as I can), so don't hold your breath. I'll ask here if I have any questions about the alda side of things (I probably will). 

I'll be working [here](https://github.com/jgkamat/alda-client-java/tree/java-repl) just for reference.

I had one more question for now though, would it be possible to drop java 7 support? I wanted to use some of the java 8 features if possible, like method references, but if you want to keep java 7 compatibility I can manage. 
(Java 7 was deprecated in 2015, so I would think that most people would have migrated to java 8, but I'm not sure how many people are stil behind).

**Comment by daveyarwood** [comment]
@jgkamat No rush! I think it's great that you're helping me with this. I've been busy lately with work and family, it would probably be a much longer time before I could get around to working on this myself. The work you've done so far has already been super helpful.

I'm totally fine with dropping Java 7 support. At this point, it's probably been long enough that we can expect most Alda users to have Java 8 installed, and we do have a link in the installation instructions to download the latest Java version, for anyone who still has Java 7.

**Comment by daveyarwood** [comment]
Almost done!

TODO:

- https://github.com/alda-lang/alda-client-java/issues/10
- https://github.com/alda-lang/alda-server-clj/issues/7
- Parse the data returned by the server when playing a score with this new option, and update the REPL prompt with the current instruments.

**Comment by daveyarwood** [comment]
I just deployed [alda-server-clj 0.1.5](https://github.com/alda-lang/alda-server-clj/commit/5a6d16796608feaf2c91a835036bafddc7922650), which adds the parsed score (as a JSON object) to the response when responding to `play-status` requests.

@jgkamat I think we're ready to implement getting the current instruments from the server response after playing each line of code in the REPL. Would you be interested in trying that out?

**Comment by jgkamat** [comment]
Nice! :tada:  :smile: I'm pretty busy right now, but I'll try to get something working this weekend (or earlier, if I can squeeze some time in).

EDIT: started work on this here: https://github.com/jgkamat/alda-client-java/tree/jay/repl-status
